optimizely/config_manager.py

```diff
@@ -0,0 +1,233 @@
+# Copyright 2019, Optimizely
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import abc
+import requests
+import threading
+import time
+from requests import codes as http_status_codes
+from requests import exceptions as requests_exceptions
+
+from . import exceptions as optimizely_exceptions
+from . import logger as optimizely_logger
+from . import project_config
+from .error_handler import NoOpErrorHandler as noop_error_handler
+from .helpers import enums
+
+
+ABC = abc.ABCMeta('ABC', (object,), {'__slots__': ()})
+
+
+class BaseConfigManager(ABC):
+    """ Base class for Optimizely's config manager. """
+
+    def __init__(self,
+                 logger=None,
+                 error_handler=None,
+                 **kwargs):
+        """ Initialize config manager.
+
+        Args:
+            logger: Provides a logger instance.
+            error_handler: Provides a handle_error method to handle exceptions.
+        """
+        self.logger = logger or optimizely_logger.adapt_logger(logger or optimizely_logger.NoOpLogger())
+        self.error_handler = error_handler or noop_error_handler
+
+    @abc.abstractmethod
+    def get_config(self):
+        """ Get config for use by optimizely.Optimizely.
+        The config should be an instance of project_config.ProjectConfig."""
+        pass
+
+
+class StaticConfigManager(BaseConfigManager):
+    """ Config manager that returns ProjectConfig based on provided datafile. """
+
+    def __init__(self,
+                 datafile=None,
+                 logger=None,
+                 error_handler=None):
+        """ Initialize config manager. Datafile has to be provided to use.
+
+        Args:
+            datafile: JSON string representing the Optimizely project.
+            logger: Provides a logger instance.
+            error_handler: Provides a handle_error method to handle exceptions.
+        """
+        super(StaticConfigManager, self).__init__(logger=logger, error_handler=error_handler)
+        self._config = None
+        if datafile:
+          self._config = project_config.ProjectConfig(datafile, self.logger, self.error_handler)
+
+    def get_config(self):
+        """ Returns instance of ProjectConfig.
+
+        Returns:
+          ProjectConfig.
+        """
+        return self._config
+
+
+class PollingConfigManager(BaseConfigManager):
+    """ Config manager that polls for the datafile and updated ProjectConfig based on an update interval. """
```

Comment:
This applies both here and for `StaticConfigManager`: should we have a docstring example of how to use this class? The tests do a good job, but I'm wondering where we will point developers for an example of how to use it in practice.

Reply:
Good question. I think we would want to have good documentation on developers.optimizely.com and/or add a usage section in README.md explaining everything about which class to use and when.

Reply:
I like both of these ideas! Should usage be included in this PR?

Reply:
Will add as a follow up.